### PR TITLE
Refactor Chakra and ChakraCore jsi::Runtime implementations into a class hierarchy.

### PR DESCRIPTION
**Pull request by hansenyy**
Since Chakra and ChakraCore has similar but slightly differnt APIs, we've relying heavily on conditional compilation: We use the macro USE_JSRT_EDGEMODE to check whether we are linking with Chakra.dll or ChakraCore.dll, causing code to be very difficult to navigate at times. Using USE_JSRT_EDGEMODE to perform this check is also fundamentally wrong, since it is the macro that toggles whether chakrart.h (header for chakra.dll) or jsrt9.h (header for legacy IE engine) in #included when one #include jsrt.h.

This PR:
- Removes checks for USE_JSRT_EDGEMODE within JSI code. We now use the (existing) CHAKRACORE macro for checking whether we are linking against Chakra or ChakraCore.
- Refactors the current Chakra and ChakraCore jsi::Runtime implementations into a class hierarchy. The base class ChakraRuntime contains the functionalities of jsi::Runtime that can be implemented by APIs available to both Chakra and ChakraCore. Functionalities that need to or should be implemented using Chakra or ChakraCore specific APIs are implemented by the derived classes ChakraRtRuntime and ChakraCoreRuntime, respectively. The name "ChakraRtRuntime" comes from the header that gets #included by jsrt.h when USE_EDGEMODE_JSRT is #defined.

Note that there are still some conditional compilation based on the CHAKRACORE macro. Most of them are in ChakraObjectRef.h/cpp and all have comments explaining why they are there. The rest are used to make sure that, when a specific engine is used, only the corresponding jsi::Runtime is compiled.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3673)

### CODE REVIEW

**Comment by hansenyy** [comment]
> struct PreparedScriptStore {

This really should be called BytecodeStore, since the buffer tryGetPreparedScript returns is just a bytecode buffer and is not a jsi::PreparedJavaScript. I have a PR planned to revamp this, so I am leaving this as is for now.

---
Refers to: vnext/JSI/Shared/ScriptStore.h:32 in 52bfa81. [](commit_id = 52bfa81352e1bbfc2b067b0f73f4167d4320af6f, deletion_comment = False)

**Comment by NickGerleman** [comment]
There are a lot of changes happening here. In the future would we split this into something a bit smaller? It's totally righteous to do things like switch existing code to use numeric_cast, changing comment styles, or to change the names of namespaces, etc., but having it all in the same review is going to make it easy to miss things.

File: vnext/Common/packages.config [hunk 1]
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="boost" version="1.68.0.0" targetFramework="native" />
```
Comment (NickGerleman) [inline vnext/Common/packages.config]: ><package id="boost" version="1.68.0.0" targetFramework="native" /> [](start = 2, length = 66)

Can we prevent this from getting out of sync with the package dependency in folly/packages.config? Does this interact with the work @JunielKatarn is doing?

**Comment by hansenyy** [comment]
Closing this because it shows up as way too large a change in code flow.

**Comment by hansenyy** [comment]
> There are a lot of changes happening here. In the future would we split this into something a bit smaller? It's totally righteous to do things like switch existing code to use numeric_cast, changing comment styles, or to change the names of namespaces, etc., but having it all in the same review is going to make it easy to miss things.

Yeah I agree. Did not realize how large this became until I saw it in code flow. Will break this down further into a several PRs.

**Pull request closed without merging**